devices/nand-imx6ull/dma.c

```diff
@@ -0,0 +1,184 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * i.MX 6ULL APBH DMA
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "dma.h"
+
+
+/* DMA registers */
+#define DMA_CTRL0     0
+#define DMA_CTRL0_SET 1
+#define DMA_CTRL0_CLR 2
+#define DMA_CTRL0_TOG 3
+#define DMA_CTRL1     4
+#define DMA_CTRL1_SET 5
+#define DMA_CTRL1_CLR 6
+#define DMA_CTRL1_TOG 7
+#define DMA_CTRL2     8
+#define DMA_CTRL2_SET 9
+#define DMA_CTRL2_CLR 10
+#define DMA_CTRL2_TOG 11
+#define DMA_CHAN      12
+#define DMA_CHAN_SET  13
+#define DMA_CHAN_CLR  14
+#define DMA_CHAN_TOG  15
+#define DMA_DEVSEL    16
+#define DMA_CURCMDAR  64
+#define DMA_NXTCMDAR  68
+#define DMA_CMD       72
+#define DMA_BAR       76
+#define DMA_SEMA      80
+#define DMA_DEBUG     84
+#define DMA_DEBUG2    88
+#define DMA_VERSION   512
+
+/* DMA channels */
+#define DMA_CHAN_OFFS 28
+
+
+static struct {
+	volatile u32 *const base;
+} dma_common = {
+	.base = (void *)0x1804000,
+};
+
+
+int dma_descsz(const dma_desc_t *desc)
+{
+	return sizeof(*desc) + ((desc->flags >> 12) & 0xf) * sizeof(desc->pio[0]);
+}
+
+
+int dma_check(dma_desc_t *desc, const dma_desc_t *fail)
+{
+	desc->flags = DMA_HOT | DMA_SENSE;
+	desc->size = 0;
+	desc->addr = (u32)fail;
+
+	return dma_descsz(desc);
+}
+
+
+int dma_terminate(dma_desc_t *desc, int err)
+{
+	desc->flags = DMA_DECSEMA | DMA_IRQCOMP | DMA_NOXFER;
+	desc->size = 0;
+	desc->addr = (u32)err;
+
+	return dma_descsz(desc);
+}
+
+
+void dma_sequence(dma_t *dma, const dma_desc_t *desc)
+{
+	if (dma->last != dma->first) {
+		dma->last->flags |= DMA_CHAIN;
+		dma->last->next = (u32)desc;
+	}
+	else {
+		dma->first = desc;
+	}
+	dma->last = desc;
+}
+
+
+void dma_finish(dma_t *dma)
+{
+	u8 *next = (u8 *)dma->last;
+	dma_desc_t *desc;
+
+	if (dma->last != dma->first) {
+		next += dma_descsz(dma->last);
+	}
+
+	desc = (dma_desc_t *)next;
+	next += dma_terminate(desc, EOK);
+	dma_sequence(dma, desc);
+}
+
+
+void dma_reset(dma_t *dma)
+{
+	dma->first = dma->buff;
+	dma->last = dma->buff;
+}
+
+
+int dma_run(const dma_t *dma, unsigned int chan)
+{
+	int comp = 0, err = 0;
+	u32 status;
+
+	/* Run DMA channel */
+	*(dma_common.base + chan * DMA_CHAN_OFFS + DMA_NXTCMDAR) = (u32)dma->first;
+	*(dma_common.base + chan * DMA_CHAN_OFFS + DMA_SEMA) = 1;
+
+	/* Wait for DMA completion */
+	do {
+		/* Completion status */
+		status = *(dma_common.base + DMA_CTRL1);
+		if (status & (1 << chan)) {
+			comp++;
+		}
+
+		/* Error status, 0: no error, 1: DMA termination, 2: AHB bus error */
+		status = *(dma_common.base + DMA_CTRL2);
+		if (status & (1 << chan)) {
+			if (status & (1 << (chan + 16))) {
+				err++;
+			}
+			err++;
+		}
+	} while ((comp == 0) && (err == 0));
+
+	/* DMA termination with completion flag is not treated as an error */
+	err -= err & comp;
```

Comment:
Clever but hacky at the same time and not MISRA compliant I'm affraid
```suggestion
	if (err == 1 && comp != 0) {
		err = 0;
	};
```

Reply:
Fixed. I used `comp == 1` instead of `comp != 0` - seems better to me.